### PR TITLE
#CX-15828 - Reusable web component list for infinite scrolling

### PR DESCRIPTION
**Pull request by nithishkumar98**
## Description
Reusable web component list for infinite scrolling and virtualisation

## Screenshots:
**Before** (If applicable):

**After**:

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentError.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    handleRetry() {
+      console.log('Retry triggered');
+      this.loadMoreItems();
+    }
+
+    render() {
+      console.log("rendering parent component--error is loading", this.isLoading);
+      return html`
+        <h2>Error scenario</h2>
+        <md-advance-list style="width:400px"
```
Comment (parsoman) [inline web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentError.ts:88]: Dont add css class inline

File: web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentError.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          @list-item-change=${this.handleListItemChange}
+          @load-more=${this.loadMoreItems}>
+          <md-spinner size="24" slot="spin-loader"></md-spinner>
+        </md-advance-list>
+        
+        ${this.isError ? html`
+          <md-fetch-error 
+            class="fetch-error"
+            messageType="error"
+            style="color: red;"
```
Comment (parsoman) [inline web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentError.ts:105]: Dont add inline styling.

File: web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentError.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            trackingId="12345"
+            trackingIdInputLabel="Tracking ID:"
+            @retry="${this.handleRetry}">
+          </md-fetch-error>
+        ` : ''}
+      `;
+    }
+
+    static get styles() {
+      return css`
```
Comment (parsoman) [inline web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentError.ts:117]: You can create a separate file for css.

File: web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentGeneric.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
+                const r = Math.random() * 16 | 0;
+                const v = c === 'x' ? r : (r & 0x3 | 0x8);
+                return v.toString(16);
+            });
+        }
+                
+
+        async loadMoreItems() {
+            console.log("load more called")
```
Comment (parsoman) [inline web-components/src/[sandbox]/examples/AdvanceList/components/ParentComponentGeneric.ts:35]: Remove this console logs.

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    @property({ type: Boolean }) isLoading = false;
+    @property({ type: String }) selectedItemId = "";
+    @property({ type: String }) value = "";
+    @property({ type: String }) ariaRoleList = "";
+    @property({ type: String }) ariaLabelList = "";
+    @property({ type: Boolean }) isError = false;
+    @queryAll("div.default-wrapper") lists?: HTMLDivElement[];
+    @query(".virtual-scroll") listContainer?: HTMLDivElement;
+    @internalProperty() page = 1;
+    @property({ type: Number }) totalRecords = 0;
```
Comment (parsoman) [inline web-components/src/components/advance-list/AdvanceList.ts:23]: Can we remove this totalRecords and try to work our logic without it.

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+    static get styles() {
+      return [reset, styles];
+    }
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+      // Clean up event listeners when the component is removed
+      this.removeEventListener("click", this.handleClick);
+      this.listContainer?.removeEventListener("keydown", this.handleKeyDown.bind(this));
```
Comment (parsoman) [inline web-components/src/components/advance-list/AdvanceList.ts:41]: Why we are using bind in the function. We should not use it. You can make the function as arrow function  which will do the same job as bind.

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    protected firstUpdated(_changedProperties: PropertyValues): void {
+      // Add keydown event listener to the list container
+      this.listContainer?.addEventListener("keydown", this.handleKeyDown.bind(this));
```
Comment (parsoman) [inline web-components/src/components/advance-list/AdvanceList.ts:46]: Why we are using bind in the function. We should not use it. You can make the function as arrow function which will do the same job as bind.

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+    handleRangeChange = (e: any) => {
+      const { last } = e;
+      this.updateSelectedState();
+      
+      // Trigger 'load-more' event when more items need to be loaded
+      if (this.items.length < this.totalRecords && last >= this.items.length - 1 && !this.isLoading && !this.isError) {
+        this.dispatchEvent(
+          new CustomEvent("load-more", {
+            detail: { page: this.page },
```
Comment (neelabhnagaich) [inline web-components/src/components/advance-list/AdvanceList.ts:193]: this load-more event will do the data fetching for new records correct ?
But what if that data loading fails, then how are we taking care of `this.page` in that case, as we are incrementing page here anyways.

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+      );
+    }
+
+    renderItem(item: any, index: number) {
+      return html`
+        <div
+          class="default-wrapper"
+          aria-setsize="${this.totalRecords}"
+          aria-posinset="${index + 1}"
+          role="option"
```
Comment (neelabhnagaich) [inline web-components/src/components/advance-list/AdvanceList.ts:219]: shouldn't this also needs to be dynamic as option role only goes with listbox and I see that for parent you have the dynamic role
Example: listitem

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+      }
+    }
+
+    protected updateSelectedState() {
+      const wrappers = Array.from(this.shadowRoot?.querySelectorAll(".default-wrapper") || []);
+      wrappers.forEach((wrapper) => {
+
+        if (wrapper.id === `${prefixId}${this.selectedItemId}`) {
+          // Update selected item styles and attributes
+          wrapper.classList.add("selected");
```
Comment (neelabhnagaich) [inline web-components/src/components/advance-list/AdvanceList.ts:66]: can be moved to a function for better structuring and readability
```
wrapper.classList.add("selected");
wrapper.setAttribute("selected", "true");
wrapper.setAttribute("aria-selected", "true");
wrapper.setAttribute("tabindex", "0");
```

**Review by neelabhnagaich — DISMISSED** [review]

File: web-components/src/components/advance-list/AdvanceList.ts [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+          id="${prefixId}${item.id}"
+          index="${index}"
+        >
+          ${item.template(item, index)}
+        </div>
+      `;
+    }
+
+    render() {
+      let ariaActiveDescendant: string = "";
```
Comment (parsoman) [inline web-components/src/components/advance-list/AdvanceList.ts:225]: We can make this logic as a function and directly call it in aria-activedescendant ?

File: web-components/webpack.config.ts [hunk 12]
```diff
@@ -200,7 +201,7 @@ const commonDist = merge(common, {
     "comp/md-link-entry": "./src/components/link/Link",
     "comp/md-list-entry": "./src/components/list/List",
     "comp/md-list-item-entry": "./src/components/list/ListItem",
-    "comp/md-loading-entry": "./src/components/loading/Loading",
+        "comp/md-loading-entry": "./src/components/loading/Loading",
```
Comment (parsoman) [inline web-components/webpack.config.ts:204]: Remove this tab change

**Review by parsoman — DISMISSED** [review]

**Review by neelabhnagaich — APPROVED** [review]

**Review by vikcool2pr — APPROVED** [review]

**Pull request merged**